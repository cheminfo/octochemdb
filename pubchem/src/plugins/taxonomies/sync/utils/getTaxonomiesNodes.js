import Debug from '../../../../utils/Debug.js';

const debug = Debug('getTaxonomiesNodes');
export function getTaxonomiesNodes(arrayBuffer, connection) {
  try {
    const decoder = new TextDecoder();
    arrayBuffer = new Uint8Array(arrayBuffer);
    let begin = 0;
    let end = 0;
    let result = {};

    while (end < arrayBuffer.length) {
      if (arrayBuffer[end] === 10) {
        const line = decoder.decode(arrayBuffer.subarray(begin, end));
        const fields = line.split(/[\t ]*\|[ \t]*/);

        if (fields.length < 2) {
          end++;
          continue;
        }

        let taxId = fields[0].replace(/[\r\n]/g, '');
        result[taxId] = fields[2];
        begin = end;
      }
      end++;
    }

    return result;
  } catch (e) {
    if (connection) {
      debug(e, { collection: 'taxonomies', connection });
    }
  }
}
