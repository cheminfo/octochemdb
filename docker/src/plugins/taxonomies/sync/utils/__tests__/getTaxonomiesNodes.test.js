import { readFileSync } from 'fs';
import { join } from 'path';

import { getTaxonomiesNodes } from '../getTaxonomiesNodes.js';

test('getTaxonomiesNodes', () => {
  const arrayBuffer = readFileSync(join(__dirname, 'data/nodesTest.dmp'));
  const result = getTaxonomiesNodes(arrayBuffer);
  expect(result).toMatchSnapshot();
});
