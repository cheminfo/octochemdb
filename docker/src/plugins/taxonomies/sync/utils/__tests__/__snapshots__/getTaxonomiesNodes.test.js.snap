// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getTaxonomiesNodes 1`] = `
{
  "1": "no rank",
  "10": "genus",
  "100": "species",
  "1001": "species",
  "1002": "species",
  "1003": "species",
  "1004": "species",
  "1005": "species",
  "1006": "species",
  "1007": "genus",
  "1008": "species",
  "101": "genus",
  "1010": "species",
  "1011": "genus",
  "1012": "species",
  "1013": "genus",
  "1014": "species",
  "1015": "species",
  "1016": "genus",
  "1017": "species",
  "1018": "species",
  "1019": "species",
  "102": "species",
  "1021": "genus",
  "1022": "species",
  "1023": "genus",
  "1025": "species",
  "1026": "species",
  "1027": "species",
  "1028": "species",
  "103": "species",
  "1030": "genus",
  "1031": "species",
  "1032": "species",
  "1033": "genus",
  "1034": "species",
  "1035": "species",
  "1037": "species",
  "104": "species",
  "1041": "genus",
  "1042": "species",
  "1044": "species",
  "1046": "family",
  "1047": "genus",
  "1048": "species",
  "1049": "species",
  "105": "genus",
  "1050": "species",
  "1051": "genus",
  "1052": "species",
  "1053": "species",
  "1054": "species",
  "1056": "genus",
  "1057": "species",
  "1058": "species",
  "106": "species",
  "1060": "genus",
  "1061": "species",
  "1062": "species",
  "1063": "species",
  "1064": "genus",
  "1066": "species",
  "1067": "species",
  "1068": "genus",
  "1069": "species",
  "107": "genus",
  "1070": "genus",
  "1071": "species",
  "1073": "genus",
  "1074": "species",
  "1075": "species",
  "1076": "species",
  "1078": "species",
  "1079": "species",
  "108": "species",
  "1080": "species",
  "1081": "genus",
  "1082": "species",
  "1083": "species",
  "1084": "species",
  "1085": "species",
  "1086": "species",
  "1087": "species",
  "1088": "species",
  "109": "genus",
  "1090": "phylum",
  "1091": "genus",
  "1092": "species",
  "1094": "species",
  "1095": "species",
  "1096": "species",
  "1097": "species",
  "1098": "species",
  "1099": "genus",
  "11": "species",
  "110": "species",
  "1100": "species",
  "1101": "genus",
  "1102": "species",
  "1103": "genus",
  "1104": "species",
  "1106": "family",
  "1107": "genus",
  "1108": "species",
  "1109": "genus",
  "1110": "species",
  "1111": "genus",
  "1112": "species",
  "1113": "species",
  "1117": "phylum",
  "1118": "order",
  "1119": "genus",
  "112": "order",
  "1121": "genus",
  "1122": "species",
  "1123": "genus",
  "1124": "species",
  "1125": "genus",
  "1126": "species",
  "1127": "species",
  "1128": "species",
  "1129": "genus",
  "113": "genus",
  "1131": "species",
  "114": "species",
  "1140": "strain",
  "1142": "genus",
  "1143": "species",
  "1144": "species",
  "1147": "species",
  "1148": "species",
  "1150": "order",
  "1151": "species",
  "1152": "genus",
  "1153": "species",
  "1154": "genus",
  "1157": "species",
  "1158": "genus",
  "1159": "species",
  "1160": "species",
  "1161": "order",
  "1162": "family",
  "1163": "genus",
  "1164": "species",
  "1165": "species",
  "1166": "species",
  "1167": "species",
  "1168": "species",
  "117": "species",
  "1173": "species",
  "1175": "genus",
  "1176": "species",
  "1177": "genus",
  "1178": "species",
  "1179": "species",
  "118": "genus",
  "1180": "species",
  "1181": "strain",
  "1182": "family",
  "1183": "genus",
  "1184": "species",
  "1185": "family",
  "1186": "genus",
  "1187": "species",
  "1188": "species",
  "1189": "family",
  "119": "species",
  "1190": "genus",
  "1191": "species",
  "1197": "species",
  "1198": "genus",
  "1199": "species",
  "120": "species",
  "1200": "species",
  "1202": "species",
  "1203": "genus",
  "1204": "species",
  "1205": "genus",
  "1206": "species",
  "1207": "species",
  "1208": "species",
  "1211": "species",
  "1213": "family",
  "1214": "genus",
  "1215": "species",
  "1216": "species",
  "1218": "genus",
  "1219": "species",
  "122": "species",
  "1220": "species",
  "1222": "genus",
  "1223": "species",
  "1224": "phylum",
  "1227": "genus",
  "1229": "species",
  "123": "genus",
  "1231": "species",
  "1232": "genus",
  "1233": "species",
  "1234": "genus",
  "1236": "class",
  "1237": "genus",
  "1238": "species",
  "1239": "phylum",
  "124": "species",
  "1243": "genus",
  "1244": "species",
  "1245": "species",
  "1246": "species",
  "1247": "species",
  "1249": "species",
  "125": "species",
  "1251": "species",
  "1252": "species",
  "1253": "genus",
  "1254": "species",
  "1255": "species",
  "1257": "genus",
  "1258": "species",
  "1259": "species",
  "126": "family",
  "1260": "species",
  "1261": "species",
  "1262": "species",
  "1263": "genus",
  "1264": "species",
  "1265": "species",
  "1266": "genus",
  "1267": "species",
  "1268": "family",
  "1269": "genus",
  "127": "genus",
  "1270": "species",
  "1271": "species",
  "1272": "species",
  "1273": "species",
  "1274": "species",
  "1275": "species",
  "1276": "species",
  "1279": "genus",
  "128": "species",
  "1280": "species",
  "1281": "species",
  "1282": "species",
  "1283": "species",
  "1284": "species",
  "1285": "species",
  "1286": "species",
  "1287": "no rank",
  "1288": "species",
  "1290": "species",
  "1292": "species",
  "1293": "species",
  "1294": "species",
  "1295": "species",
  "1296": "species",
  "1297": "phylum",
  "1298": "genus",
  "1299": "species",
  "13": "genus",
  "131": "genus",
  "132": "species",
  "133": "genus",
  "134": "species",
  "136": "order",
  "137": "family",
  "138": "genus",
  "139": "species",
  "14": "species",
  "140": "species",
  "141": "species",
  "142": "species",
  "143": "species",
  "144": "species",
  "145": "species",
  "146": "genus",
  "147": "species",
  "148": "species",
  "149": "species",
  "150": "species",
  "151": "species",
  "152": "species",
  "154": "species",
  "156": "species",
  "157": "genus",
  "158": "species",
  "159": "species",
  "16": "genus",
  "160": "species",
  "161": "subspecies",
  "162": "species",
  "163": "species",
  "164": "species",
  "165": "species",
  "166": "species",
  "167": "species",
  "168": "subspecies",
  "17": "species",
  "170": "family",
  "171": "genus",
  "172": "species",
  "173": "species",
  "174": "species",
  "176": "no rank",
  "178": "species",
  "179": "genus",
  "18": "genus",
  "180": "species",
  "181": "species",
  "183": "species",
  "185": "species",
  "187": "genus",
  "188": "species",
  "189": "species",
  "19": "species",
  "190": "species",
  "191": "genus",
  "192": "species",
  "193": "species",
  "194": "genus",
  "195": "species",
  "196": "species",
  "197": "species",
  "198": "species",
  "199": "species",
  "2": "superkingdom",
  "20": "genus",
  "200": "species",
  "201": "species",
  "202": "species",
  "203": "species",
  "204": "species",
  "205": "species",
  "206": "species",
  "207": "species",
  "209": "genus",
  "21": "species",
  "210": "species",
  "211": "species",
  "212": "species",
  "213": "species",
  "214": "species",
  "215": "species",
  "216": "species",
  "217": "species",
  "218": "species",
  "22": "genus",
  "222": "genus",
  "223": "species",
  "224": "species",
  "226": "genus",
  "227": "species",
  "228": "species",
  "23": "species",
  "232": "species",
  "234": "genus",
  "235": "species",
  "236": "species",
  "237": "genus",
  "238": "species",
  "239": "species",
  "24": "species",
  "240": "species",
  "241": "species",
  "242": "species",
  "244": "species",
  "245": "species",
  "246": "species",
  "247": "species",
  "249": "species",
  "25": "species",
  "250": "species",
  "251": "species",
  "252": "species",
  "253": "species",
  "254": "species",
  "255": "species",
  "256": "species",
  "258": "species",
  "259": "species",
  "262": "genus",
  "263": "species",
  "264": "subspecies",
  "265": "genus",
  "266": "species",
  "267": "species",
  "27": "species",
  "270": "genus",
  "271": "species",
  "272": "species",
  "274": "species",
  "275": "species",
  "276": "species",
  "277": "species",
  "279": "genus",
  "28": "species",
  "280": "species",
  "281": "species",
  "283": "genus",
  "285": "species",
  "286": "genus",
  "287": "species",
  "288": "species",
  "29": "order",
  "290": "species",
  "292": "species",
  "293": "species",
  "294": "species",
  "296": "species",
  "297": "species",
  "298": "species",
  "300": "species",
  "301": "species",
  "303": "species",
  "304": "species",
  "305": "species",
  "306": "species",
  "307": "species",
  "308": "species",
  "309": "species",
  "31": "family",
  "311": "species",
  "312": "species",
  "313": "species",
  "314": "species",
  "315": "species",
  "316": "species",
  "317": "species",
  "318": "no rank",
  "319": "no rank",
  "32": "genus",
  "321": "no rank",
  "322": "no rank",
  "323": "no rank",
  "329": "species",
  "33": "species",
  "333": "subspecies",
  "335": "no rank",
  "337": "species",
  "338": "genus",
  "339": "species",
  "34": "species",
  "340": "no rank",
  "343": "species",
  "346": "species",
  "347": "species",
  "349": "genus",
  "35": "species",
  "350": "species",
  "351": "subfamily",
  "352": "genus",
  "353": "species",
  "354": "species",
  "355": "strain",
  "356": "order",
  "357": "genus",
  "358": "species",
  "359": "species",
  "361": "species",
  "370": "strain",
  "371": "strain",
  "372": "strain",
  "373": "species",
  "374": "genus",
  "375": "species",
  "376": "species",
  "378": "species",
  "379": "genus",
  "38": "species",
  "380": "species",
  "381": "species",
  "382": "species",
  "384": "species",
  "385": "no rank",
  "386": "no rank",
  "387": "no rank",
  "39": "family",
  "391": "species",
  "393": "species",
  "394": "strain",
  "396": "species",
  "398": "species",
  "399": "species",
  "40": "genus",
  "400": "species",
  "403": "family",
  "404": "genus",
  "405": "species",
  "406": "species",
  "407": "genus",
  "408": "species",
  "409": "species",
  "41": "species",
  "410": "species",
  "411": "species",
  "413": "genus",
  "414": "species",
  "415": "species",
  "416": "genus",
  "417": "species",
  "418": "species",
  "42": "genus",
  "421": "species",
  "423": "species",
  "424": "species",
  "425": "genus",
  "426": "species",
  "427": "species",
  "428": "species",
  "429": "genus",
  "43": "species",
  "430": "species",
  "433": "family",
  "434": "genus",
  "435": "species",
  "436": "species",
  "437": "species",
  "438": "species",
  "439": "species",
  "44": "genus",
  "440": "species",
  "441": "genus",
  "442": "species",
  "444": "family",
  "445": "genus",
  "446": "species",
  "447": "species",
  "448": "species",
  "449": "species",
  "45": "species",
  "450": "species",
  "451": "species",
  "452": "species",
  "453": "species",
  "454": "species",
  "455": "species",
  "456": "species",
  "458": "species",
  "459": "species",
  "460": "species",
  "461": "genus",
  "463": "species",
  "464": "species",
  "465": "genus",
  "466": "species",
  "468": "family",
  "469": "genus",
  "47": "genus",
  "470": "species",
  "471": "species",
  "472": "species",
  "474": "species",
  "475": "genus",
  "476": "species",
  "477": "species",
  "478": "species",
  "479": "species",
  "48": "species",
  "480": "species",
  "481": "family",
  "482": "genus",
  "483": "species",
  "484": "species",
  "485": "species",
  "486": "species",
  "487": "species",
  "488": "species",
  "489": "species",
  "49": "family",
  "490": "species",
  "491": "serogroup",
  "492": "species",
  "493": "species",
  "494": "species",
  "495": "species",
  "496": "species",
  "497": "genus",
  "498": "species",
  "499": "genus",
  "50": "genus",
  "500": "species",
  "502": "species",
  "504": "species",
  "505": "species",
  "506": "family",
  "507": "genus",
  "51": "species",
  "511": "species",
  "512": "species",
  "516": "species",
  "517": "genus",
  "518": "species",
  "519": "species",
  "52": "species",
  "520": "species",
  "521": "species",
  "522": "genus",
  "523": "species",
  "524": "species",
  "525": "species",
  "526": "species",
  "527": "species",
  "528": "genus",
  "529": "species",
  "53": "genus",
  "531": "species",
  "532": "genus",
  "533": "species",
  "535": "genus",
  "536": "species",
  "537": "species",
  "538": "genus",
  "539": "species",
  "54": "species",
  "540": "species",
  "541": "genus",
  "542": "species",
  "543": "family",
  "544": "genus",
  "545": "species",
  "546": "species",
  "547": "genus",
  "548": "species",
  "549": "species",
  "55": "genus",
  "550": "species",
  "551": "genus",
  "552": "species",
  "553": "species",
  "554": "species",
  "555": "subspecies",
  "556": "species",
  "558": "species",
  "56": "species",
  "561": "genus",
  "562": "species",
  "563": "species",
  "564": "species",
  "565": "species",
  "566": "species",
  "568": "genus",
  "569": "species",
  "57": "species",
  "570": "genus",
  "571": "species",
  "573": "species",
  "574": "subspecies",
  "575": "species",
  "576": "species",
  "577": "species",
  "579": "genus",
  "580": "species",
  "581": "genus",
  "582": "species",
  "583": "genus",
  "584": "species",
  "585": "species",
  "586": "genus",
  "587": "species",
  "588": "species",
  "589": "species",
  "59": "genus",
  "590": "genus",
  "594": "no rank",
  "595": "no rank",
  "596": "no rank",
  "597": "no rank",
  "598": "no rank",
  "599": "species",
  "6": "genus",
  "60": "species",
  "600": "no rank",
  "604": "no rank",
  "605": "no rank",
  "607": "no rank",
  "61": "species",
  "610": "no rank",
  "611": "no rank",
  "612": "no rank",
  "613": "genus",
  "614": "species",
  "615": "species",
  "616": "species",
  "617": "species",
  "618": "species",
  "62": "species",
  "620": "genus",
  "621": "species",
  "622": "species",
  "623": "species",
  "624": "species",
  "625": "species",
  "626": "genus",
  "628": "species",
  "629": "genus",
  "63": "species",
  "630": "species",
  "631": "species",
  "632": "species",
  "633": "species",
  "634": "species",
  "635": "genus",
  "636": "species",
  "637": "genus",
  "638": "species",
  "64": "genus",
  "641": "family",
  "642": "genus",
  "644": "species",
  "645": "species",
  "646": "species",
  "647": "species",
  "648": "species",
  "649": "species",
  "65": "species",
  "650": "species",
  "651": "species",
  "652": "species",
  "654": "species",
  "656": "species",
  "657": "genus",
  "658": "subspecies",
  "659": "species",
  "660": "species",
  "661": "species",
  "662": "genus",
  "663": "species",
  "664": "no rank",
  "666": "species",
  "668": "species",
  "669": "species",
  "670": "species",
  "671": "species",
  "672": "species",
  "673": "species",
  "674": "species",
  "675": "species",
  "676": "species",
  "678": "species",
  "679": "species",
  "68": "genus",
  "680": "species",
  "683": "species",
  "685": "species",
  "686": "no rank",
  "687": "species",
  "688": "species",
  "689": "species",
  "69": "species",
  "691": "species",
  "693": "species",
  "7": "species",
  "701": "species",
  "702": "genus",
  "703": "species",
  "704": "species",
  "709": "species",
  "71": "genus",
  "712": "family",
  "713": "genus",
  "714": "species",
  "715": "species",
  "716": "species",
  "717": "species",
  "718": "species",
  "719": "species",
  "72": "species",
  "720": "species",
  "721": "species",
  "722": "species",
  "723": "species",
  "724": "genus",
  "725": "biotype",
  "726": "species",
  "727": "species",
  "728": "species",
  "729": "species",
  "730": "species",
  "731": "species",
  "732": "species",
  "733": "species",
  "734": "species",
  "735": "species",
  "736": "species",
  "738": "species",
  "739": "species",
  "740": "species",
  "741": "species",
  "745": "genus",
  "747": "species",
  "749": "species",
  "75": "genus",
  "750": "species",
  "751": "species",
  "752": "species",
  "753": "species",
  "754": "species",
  "755": "species",
  "756": "species",
  "757": "species",
  "758": "species",
  "759": "species",
  "760": "species",
  "761": "species",
  "762": "species",
  "766": "order",
  "768": "genus",
  "769": "species",
  "770": "species",
  "772": "family",
  "773": "genus",
  "774": "species",
  "775": "family",
  "776": "genus",
  "777": "species",
  "779": "species",
  "78": "species",
  "780": "genus",
  "781": "species",
  "782": "species",
  "783": "species",
  "784": "species",
  "785": "species",
  "786": "species",
  "787": "species",
  "788": "species",
  "789": "species",
  "801": "genus",
  "803": "species",
  "807": "species",
  "809": "family",
  "81": "genus",
  "810": "genus",
  "813": "species",
  "815": "family",
  "816": "genus",
  "817": "species",
  "818": "species",
  "82": "species",
  "820": "species",
  "821": "species",
  "823": "species",
  "824": "species",
  "826": "species",
  "827": "species",
  "83": "species",
  "830": "genus",
  "831": "species",
  "832": "genus",
  "833": "species",
  "834": "subspecies",
  "836": "genus",
  "837": "species",
  "838": "genus",
  "839": "species",
  "84": "species",
  "840": "species",
  "841": "genus",
  "842": "species",
  "843": "genus",
  "844": "species",
  "846": "genus",
  "847": "species",
  "848": "genus",
  "849": "species",
  "85": "genus",
  "850": "species",
  "851": "species",
  "852": "species",
  "853": "species",
  "854": "species",
  "855": "species",
  "856": "species",
  "858": "species",
  "859": "species",
  "86": "species",
  "860": "species",
  "861": "species",
  "862": "genus",
  "863": "species",
  "864": "genus",
  "865": "species",
  "866": "genus",
  "867": "species",
  "868": "family",
  "869": "genus",
  "87": "species",
  "870": "species",
  "872": "genus",
  "873": "species",
  "876": "species",
  "878": "species",
  "879": "species",
  "88": "genus",
  "880": "species",
  "881": "species",
  "882": "strain",
  "883": "strain",
  "884": "strain",
  "885": "species",
  "886": "species",
  "888": "species",
  "889": "species",
  "89": "species",
  "890": "genus",
  "891": "species",
  "892": "species",
  "893": "genus",
  "894": "species",
  "895": "species",
  "896": "genus",
  "897": "species",
  "898": "genus",
  "899": "species",
  "9": "species",
  "901": "species",
  "904": "genus",
  "905": "species",
  "906": "genus",
  "907": "species",
  "911": "genus",
  "912": "species",
  "913": "species",
  "914": "genus",
  "915": "species",
  "916": "species",
  "917": "species",
  "919": "genus",
  "920": "species",
  "921": "species",
  "923": "species",
  "924": "species",
  "926": "species",
  "927": "species",
  "928": "species",
  "929": "species",
  "93": "genus",
  "930": "species",
  "931": "species",
  "932": "species",
  "933": "genus",
  "934": "species",
  "935": "species",
  "936": "genus",
  "937": "species",
  "939": "genus",
  "94": "species",
  "940": "species",
  "942": "family",
  "943": "genus",
  "944": "species",
  "945": "species",
  "947": "species",
  "948": "species",
  "949": "species",
  "950": "species",
  "951": "species",
  "952": "tribe",
  "953": "genus",
  "954": "species",
  "955": "species",
  "956": "species",
  "958": "genus",
  "959": "species",
  "96": "genus",
  "960": "species",
  "962": "species",
  "963": "genus",
  "964": "species",
  "965": "genus",
  "966": "species",
  "967": "genus",
  "968": "species",
  "97": "species",
  "970": "genus",
  "971": "species",
  "972": "family",
  "973": "genus",
  "974": "species",
  "976": "phylum",
  "978": "genus",
  "979": "species",
  "981": "species",
  "982": "species",
  "983": "species",
  "984": "species",
  "985": "species",
  "986": "species",
  "987": "species",
  "988": "species",
  "989": "species",
  "99": "genus",
  "990": "species",
  "991": "species",
  "992": "genus",
  "995": "species",
  "996": "species",
  "997": "species",
  "998": "species",
  "999": "species",
}
`;
